### PR TITLE
Adding L2 and Dropout regularization to CQL loss

### PR DESCRIPTION
**Pull request by abhiksingla**
## Why are these changes needed?
Regularization is required to solve the inherent issue of overfitting in the CQL algorithm. The L2 and dropout regularization stand as baselines. Additionally, added support for configuring alpha limits from the trainer config.

1. The L2 regularizer will be enabled via trainer config using the field: "enable_l2_regularizer".
2. If regularizer is enabled, gamma value can be provided using trainer config and field: "gamma"
3. The dropout will be enabled via trainer config using the field "dropout_rate". If the field is set positive (0.0 < dropout_rate <= 1.0) it will be enabled, otherwise disabled by default.
4. Upper and lower limits of lagrangian constraint for alpha value can be set using trainer config using fields: "alpha_upper_bound" and "alpha_lower_bound".

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/cql/cql_sac.py [hunk 1]
```diff
@@ -31,14 +31,24 @@
...
         # training
         "buffer_size": 1000000,
+        # Enable L2 regularization in the Q-function
+        "l2_regularizer": False,
+        # Scaling factor gamma for L2 regularization
+        "l2_gamma": 100,
+        # Dropout rate
+        "dropout_rate": 0.0,
+        # Upper bound for alpha value during the langrangian constraint
+        "alpha_upper_bound": 1.0,
```
Comment (Edilmo) [inline rllib/agents/cql/cql_sac.py:43]: I would use a reasonable value for both, these values only apply if lagrangian is enabled.
There is a flag, lagrangian, which is already in False in RLlib, but it is True in bonsai. If this is False, the lagrangian is disabled. which is what you are actually looking for.

**Review by Edilmo — APPROVED** [review]
After addressing the comment you could land this

File: rllib/agents/sac/sac_tf_model.py [hunk 2]
```diff
@@ -125,18 +127,24 @@ def build_action_model(self):
...
-        q_net = tf.keras.Sequential(([
-            tf.keras.layers.Concatenate(axis=1),
-        ] if not self.discrete else []) + [
-            tf.keras.layers.Dense(
-                units=units,
-                activation=getattr(tf.nn, self.critic_hidden_activation[i], None),
-                name="{}_hidden_{}".format(name, i))
-            for i, units in enumerate(self.critic_hiddens)
-        ] + [
+        q_net = [tf.keras.layers.Concatenate(axis=1)] if not self.discrete else []
```
Comment (Edilmo) [inline rllib/agents/sac/sac_tf_model.py]: We can not make the change this way.

Let's keep the code path as it was and make and if based on the parameter, to take another code path with dropout.

There will be code repetition but that is better to facilitate the maintenance with upstream and avoid regressions.

File: rllib/agents/cql/cql_sac.py [hunk 3]
```diff
@@ -31,14 +31,24 @@
...
         "learning_starts": 1500,
         # Replay Buffer should be size of offline dataset for fastest
         # training
         "buffer_size": 1000000,
+        # Enable L2 regularization in the Q-function
+        "l2_regularizer": False,
+        # Scaling factor gamma for L2 regularization
+        "l2_gamma": 100,
+        # Dropout rate
+        "dropout_rate": 0.0,
```
Comment (Edilmo) [inline rllib/agents/cql/cql_sac.py]: This parameter should be in sac and not here. Or in both. But should be in sac because the actual change is in that agent.

File: rllib/agents/sac/sac_tf_model.py (hunk 2 above)
Comment (Edilmo) [reply rllib/agents/sac/sac_tf_model.py]: You could also make the change in place as you did with the if there, but trying to Kepler the code as close as possible to what it was. In other words keeping the list comprehension. That way the change will be obvious to any maintainer.

**Review by Edilmo — APPROVED** [review]
Good job
